android/src/main/java/com/tailscale/ipn/App.java

```diff
@@ -87,40 +87,28 @@ public class App extends Application {
 	public DnsConfig dns = new DnsConfig(this);
 	public DnsConfig getDnsConfigObj() { return this.dns; }
 
+	private ConnectivityManager connectivityManager;
+	private ConnectivityManager.NetworkCallback dnsCallback;
+	private ConnectivityManager.NetworkCallback connectivityCallback;
+
 	@Override public void onCreate() {
 		super.onCreate();
 		// Load and initialize the Go library.
 		Gio.init(this);
-		registerNetworkCallback();
+
+		this.connectivityManager = (ConnectivityManager) this.getSystemService(Context.CONNECTIVITY_SERVICE);
+		setAndRegisterNetworkCallbacks();
 
 		createNotificationChannel(NOTIFY_CHANNEL_ID, "Notifications", NotificationManagerCompat.IMPORTANCE_DEFAULT);
 		createNotificationChannel(STATUS_CHANNEL_ID, "VPN Status", NotificationManagerCompat.IMPORTANCE_LOW);
 		createNotificationChannel(FILE_CHANNEL_ID, "File transfers", NotificationManagerCompat.IMPORTANCE_DEFAULT);
 
 	}
 
-	private void registerNetworkCallback() {
-		ConnectivityManager cMgr = (ConnectivityManager) this.getSystemService(Context.CONNECTIVITY_SERVICE);
-		cMgr.registerNetworkCallback(new NetworkRequest.Builder().build(), new ConnectivityManager.NetworkCallback() {
-			private void reportConnectivityChange() {
-				NetworkInfo active = cMgr.getActiveNetworkInfo();
-				// https://developer.android.com/training/monitoring-device-state/connectivity-status-type
-				boolean isConnected = active != null && active.isConnectedOrConnecting();
-				onConnectivityChanged(isConnected);
-			}
 
-			@Override
-			public void onLost(Network network) {
-				super.onLost(network);
-				this.reportConnectivityChange();
-			}
-
-			@Override
-			public void onLinkPropertiesChanged(Network network, LinkProperties linkProperties) {
-				super.onLinkPropertiesChanged(network, linkProperties);
-				this.reportConnectivityChange();
-			}
-		});
+	private void setAndRegisterNetworkCallbacks() {
+		dnsCallback = dns.getDnsConfigCallback(connectivityManager);
+		connectivityManager.requestNetwork(dns.getDNSConfigNetworkRequest(), dnsCallback);
```

Comment:
I'm wondering, now that the new callback has replaced the old one: should it live here on the App object and just push an update into the DnsConfig object to capture new DNS settings? It's a bit odd to have a callback on DnsConfig actually be responsible for all network change notifications, not just DNS.

So, like, have an UpdateDNSFromNetwork(Network net) method on DnsConfig, move the networkrequest callback back to App, and have the callback in App call UpdateDNSFromNetwork as needed, before calling the native function?

Reply:
Done.